### PR TITLE
Document and test order and stability properties of lax.top_k.

### LINKED ISSUE #27594: Clarify whether `lax.top_k` returns results in sorted order
**Feature request:** Clarify in the documentation for [`lax.top_k`](https://docs.jax.dev/en/latest/_autosummary/jax.lax.top_k.html) whether it is guaranteed to return results in sorted order.

### PR DESCRIPTION
**Pull request by carlosgmartin**
Document and test the order and stability properties of [`lax.top_k`](https://docs.jax.dev/en/latest/_autosummary/jax.lax.top_k.html).

Fixes https://github.com/jax-ml/jax/issues/27594.

### CODE REVIEW

File: tests/lax_test.py [hunk 1]
```diff
@@ -2627,19 +2627,17 @@ def args_maker():
...
   )
-  def testTopK(self, shape, dtype, k, negative):
+  def testTopK(self, shape, dtype, k):
     def args_maker():
-      flat_values = np.arange(math.prod(shape), dtype=dtype)
-      values = self.rng().permutation(flat_values).reshape(shape)
-      if negative:
-        values = -values
-      return [values]
+      return [self.rng().randint(20, size=shape).astype(dtype)]
```
Comment (jakevdp) [inline tests/lax_test.py]: Lets use the same pattern used elsewhere: define `rng = jtu.rand_some_equal(self.rng())` outside the args maker and then `return [rng(shape, dtype)]` here.

File: tests/lax_test.py [hunk 2]
```diff
@@ -2627,19 +2627,17 @@ def args_maker():
...
     def reference_top_k(x):
-      bcast_idxs = np.broadcast_to(np.arange(shape[-1], dtype=np.int32), shape)
-      sorted_vals, sorted_idxs = lax_reference.sort_key_val(x, bcast_idxs)
-      return sorted_vals[..., :-k-1:-1], sorted_idxs[..., :-k-1:-1]
+      indices = x.shape[-1] - 1 - np.argsort(x[..., ::-1], stable=True)[..., ::-1]
+      values = np.take_along_axis(x, indices)
+      indices = indices[..., :k]
+      values = values[..., :k]
+      return values, indices
```
Comment (jakevdp) [inline tests/lax_test.py]: Let's move this definition to `lax_reference.py` and then call `lax_reference.top_k` directly.

**Review by jakevdp — CHANGES_REQUESTED** [review]

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

File: jax/_src/lax_reference.py [hunk 3]
```diff
@@ -528,3 +528,10 @@ def reducer(operand, axis=0):
       result[out_idx] = py_binop(result[out_idx], operand[idx])
     return result
   return reducer
+
+def top_k(operand, k):
+  indices = operand.shape[-1] - 1 - np.argsort(operand[..., ::-1], stable=True)[..., ::-1]
```
Comment (jakevdp) [inline jax/_src/lax_reference.py]: I think `stable=True` was added in NumPy 2.0. Probably the best would be to branch here on NumPy version and use the older syntax (I think it was `kind='stable'`?)

Comment (carlosgmartin) [reply jax/_src/lax_reference.py]: Also looks like another error was caused because older versions of NumPy required an axis argument for `numpy.take_along_axis`, as stated [here](https://numpy.org/doc/stable/release/2.3.0-notes.html):

> The parameter `axis` in `numpy.take_along_axis` function has now a default value of `-1`. ([gh-28615](https://github.com/numpy/numpy/pull/28615))

I've fixed that by explicitly setting `axis=-1`.

Comment (jakevdp) [reply jax/_src/lax_reference.py]: Actually, it looks like `kind='stable'` is valid in all NumPy versions, so you could just switch to that.

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

**Comment by carlosgmartin** [comment]
Fixed an error that occurs because `numpy.argsort` outputs int32 while `jax.lax.top_k` outputs int64 for the indices.

**Comment by carlosgmartin** [comment]
@jakevdp Minor point, but I didn't see [this](https://github.com/jax-ml/jax/pull/30995#issuecomment-3184996052) error when I ran tests locally, and only saw it when the tests ran on GitHub. This is because it only happened in x64 mode. 

It might be helpful to add a reminder to the [contributing page](https://docs.jax.dev/en/latest/contributing.html) to run tests with `JAX_ENABLE_X64=True` as well. WDYT?

**Comment by jakevdp** [comment]
Sure, sounds good!

**Comment by carlosgmartin** [comment]
Done: https://github.com/jax-ml/jax/pull/31034.
